NLSY/nlsy_lib.R

```diff
@@ -248,6 +251,191 @@ nlsy_get_highest_grade_completed_df = function()
+        pivot_longer(-c('id', 'income_source'),
+                     names_to = 'year', values_to = 'income') |>
+        mutate(year = as.numeric(str_sub(year, start=-4)))
+    networth_97_03 <- nlsydf |>
+        select(id = PUBID_1997, matches('CV_HH_NET_WORTH_Y')) |>
+        pivot_longer(-id,
+                     names_to = 'year', values_to = 'networth') |>
+        mutate(year = as.numeric(str_sub(year, start=-4)))
+
+    networth_5yr <- nlsydf |>
+        select(id = PUBID_1997, age_resp = "CV_AGE_12/31/96_1997", matches('CVC_HH_NET_WORTH_\\d{2}')) |>
+        pivot_longer(-c('id', 'age_resp'),
+                     names_to = 'year', values_to = 'networth') |>
+        mutate(year = as.numeric(str_sub(year, start=-7, end=-6)), # Get age for 5-yr wealth vars
+               year = 1997 + (year - age_resp)) # Add correct number of years based on resp's starting age
+
+    networth_merged <- tidylog::full_join(networth_97_03, networth_5yr, by = c('id', 'year', 'networth')) |>
+        select(-age_resp)
+
+    dups <- count(networth_merged, id, year) |> filter(n>1)
+    dups_rm_na <- networth_merged |> inner_join(dups, by=c('id', 'year')) |> filter(!is.na(networth)) |> select(-n)
+    networth_merged2 <- networth_merged |>
+        anti_join(dups, by=c('id', 'year')) |> # Remove duplicate id-years
+        bind_rows(dups_rm_na) |> # Add back in the de-duplicated id-years
+        arrange(id, year)
+
+
+    income_networth <- tidylog::full_join(income, networth_merged2, by = c('id', 'year')) |>
+        group_by(id) |>
+        complete(year = 1997:2019) |>
+        filter(year <= 2019) |>
+        ungroup()
+
+    return(income_networth)
+
+}
+
+
+
+#' Interpolates income and wealth
+#' @param data: data frame with income, wealth by ID and year
+#' @param interp_type: Type of interpolation, either approx for linear or spline for nonlinear
+nlsy_interp_inc_nw <- function(data, interp_type=approx){
+    # Can't interpolate for these IDs
+    missing_ids_income <- data |> summarize(income_na = sum(is.na(income)), .by='id') |> filter(income_na >= 22) |> pull(id)
+    missing_ids_nw <- data |> summarize(networth_na = sum(is.na(networth)), .by='id') |> filter(networth_na >= 22) |> pull(id)
+
+    interp_income_nw <- full_join(
+        data |>
+            filter(!(id %in% missing_ids_income)) |>
+            group_by(id) |>
+            do(grouped_interpolate(., variable='income', interp_type=interp_type)) |>
+            ungroup(),
+        data |>
+            filter(!(id %in% missing_ids_nw)) |>
+            group_by(id) |>
+            do(grouped_interpolate(., variable='networth', interp_type=interp_type)) |>
+            ungroup(),
+        by = c('id', 'year')
+    )
+    return(interp_income_nw)
+}
+
+
+#' Extrapolates income and wealth
+#' #' @param data: data frame with income, wealth interpolated by ID and year
+nlsy_extrap_inc_nw <- function(data){
```

Comment:
Not sure if we want to combine interp/extrap into one function? I also did not yet add it to make_spell because I figured there might be further tweaks